File: pys/commucation.py
import sys
import socket
from enum import Enum


class TYPE(Enum):
	Ariticle = 0
	Definition = 1
	Property = 2
	Example = 3
	Exercise = 4


class CommProtocol(object):

	def __init__(self, type_or_raw, content = None):
		if content is None:
			self.type = int(type_or_raw[0])
			self.content = type_or_raw[1:]
		else:
			self.type = type_or_raw
			self.content = content

	def to_string(self):
		return str(self.type) + str(self.content)


class Connection(object):

	@staticmethod
	def get_result(cls, type_, content):
		sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
		sock.connect("/tmp/datastructureexpt.sock")
		sock.sendall(CommProtocol(type_, content).to_string())

		raw = ""
		while True:
			tmp = sock.recv(4096)
			if tmp: raw += tmp
			else: break

		return CommProtocol(raw)

File: src/commucation.cpp
#include "commucation.hpp"


CommProtocol::CommProtocol(const std::string &raw) : type_(static_cast<TYPE>(raw[0] - '0')), content_(raw.substr(1))
{
    // ...
}

CommProtocol::CommProtocol(TYPE type, const std::string &content) : type_(type), content_(content)
{
    // ...
}

CommProtocol::TYPE CommProtocol::type() const
{
    return type_;
}

std::string CommProtocol::content() const
{
    return content_;
}

std::string CommProtocol::to_string() const
{
    return std::to_string(type_) + content_;
}

std::string CommProtocol::to_string(const CommProtocol &commp)
{
    return commp.to_string();
}

std::string CommProtocol::to_string(TYPE type, const std::string &content)
{
    return CommProtocol(type, content).to_string();
}
